File: LogisimPlus/LogisimPlus/logisim/data/BitWidth.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.data
{

	using StringGetter = logisim.util.StringGetter;

	public class BitWidth : IComparable<BitWidth>
	{
		public static readonly BitWidth UNKNOWN = new BitWidth(0);
		public static readonly BitWidth ONE = new BitWidth(1);

		private static BitWidth[] prefab = null;

		internal class Attribute : logisim.data.Attribute<BitWidth>
		{
			internal BitWidth[] choices;

			public Attribute(string name, StringGetter disp) : base(name, disp)
			{
				ensurePrefab();
				choices = prefab;
			}

			public Attribute(string name, StringGetter disp, int min, int max) : base(name, disp)
			{
				choices = new BitWidth[max - min + 1];
				for (int i = 0; i < choices.Length; i++)
				{
					choices[i] = BitWidth.create(min + i);
				}
			}

			public override BitWidth parse(string value)
			{
				return BitWidth.parse(value);
			}

			public override java.awt.Component getCellEditor(BitWidth value)
			{
				JComboBox<BitWidth> combo = new JComboBox<BitWidth>(choices);
				if (value != null)
				{
					int wid = value.Width;
					if (wid <= 0 || wid > prefab.Length)
					{
						combo.addItem(value);
					}
					combo.setSelectedItem(value);
				}
				return combo;
			}
		}

		internal readonly int width;

		private BitWidth(int width)
		{
			this.width = width;
		}

		public virtual int Width
		{
			get
			{
				return width;
			}
		}

		public virtual int Mask
		{
			get
			{
				if (width == 0)
				{
					return 0;
				}
				else if (width == 32)
				{
					return -1;
				}
				else
				{
					return (1 << width) - 1;
				}
			}
		}

		public override bool Equals(object other_obj)
		{
			if (!(other_obj is BitWidth))
			{
				return false;
			}
			BitWidth other = (BitWidth) other_obj;
			return this.width == other.width;
		}

		public virtual int CompareTo(BitWidth other)
		{
			return this.width - other.width;
		}

		public override int GetHashCode()
		{
			return width;
		}

		public override string ToString()
		{
			return "" + width;
		}

		public static BitWidth create(int width)
		{
			ensurePrefab();
			if (width <= 0)
			{
				if (width == 0)
				{
					return UNKNOWN;
				}
				else
				{
					throw new System.ArgumentException("width " + width + " must be positive");
				}
			}
			else if (width - 1 < prefab.Length)
			{
				return prefab[width - 1];
			}
			else
			{
				return new BitWidth(width);
			}
		}

		public static BitWidth parse(string str)
		{
			if (string.ReferenceEquals(str, null) || str.Length == 0)
			{
				throw new System.FormatException("Width string cannot be null");
			}
			if (str[0] == '/')
			{
				str = str.Substring(1);
			}
			return create(int.Parse(str));
		}

		private static void ensurePrefab()
		{
			if (prefab == null)
			{
				prefab = new BitWidth[Math.Min(32, Value.MAX_WIDTH)];
				prefab[0] = ONE;
				for (int i = 1; i < prefab.Length; i++)
				{
					prefab[i] = new BitWidth(i + 1);
				}
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/data/BitWidth.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.data;

import javax.swing.JComboBox;

import logisim.data.WireValue.WireValue;
import logisim.util.StringGetter;

import java.awt.Component;

public class BitWidth implements Comparable<BitWidth> {
	public static final BitWidth UNKNOWN = new BitWidth(0);
	public static final BitWidth ONE = new BitWidth(1);

	private static BitWidth[] prefab;

	static class Attribute extends logisim.data.Attribute<BitWidth> {
		private BitWidth[] choices;

		public Attribute(String name, StringGetter disp) {
			super(name, disp);
			ensurePrefab();
			choices = prefab;
		}

		public Attribute(String name, StringGetter disp, int min, int max) {
			super(name, disp);
			choices = new BitWidth[max - min + 1];
			for (int i = 0; i < choices.length; i++) choices[i] = BitWidth.create(min + i);
		}

		@Override
		public BitWidth parse(String value) {
			return BitWidth.parse(value);
		}

		@Override
		public Component getCellEditor(BitWidth value) {
			JComboBox<BitWidth> combo = new JComboBox<>(choices);
			if (value != null) {
				int wid = value.getWidth();
				if (wid <= 0 || wid > prefab.length) combo.addItem(value);
				combo.setSelectedItem(value);
			}
			return combo;
		}
	}

	final int width;

	private BitWidth(int width) {
		this.width = width;
	}

	public int getWidth() {
		return width;
	}

	public int getMask() {
		if (width == 0)
			return 0;
		else if (width == 32)
			return -1;
		else
			return (1 << width) - 1;
	}

	@Override
	public boolean equals(Object other_obj) {
		return other_obj instanceof BitWidth other && width == other.width;
	}

	public int compareTo(BitWidth other) {
		return width - other.width;
	}

	@Override
	public int hashCode() {
		return width;
	}

	@Override
	public String toString() {
		return "" + width;
	}

	public static BitWidth create(int width) {
		ensurePrefab();
		if (width <= 0) if (width == 0) return UNKNOWN;
		else throw new IllegalArgumentException("width " + width + " must be positive");
		else if (width - 1 < prefab.length) return prefab[width - 1];
		else return new BitWidth(width);
	}

	public static BitWidth parse(String str) {
		if (str == null || str.isEmpty()) throw new NumberFormatException("Width string cannot be null");
		if (str.charAt(0) == '/')
			str = str.substring(1);
		return create(Integer.parseInt(str));
	}

	private static void ensurePrefab() {
		if (prefab == null) {
			prefab = new BitWidth[Math.min(32, WireValue.MAX_WIDTH)];
			prefab[0] = ONE;
			for (int i = 1; i < prefab.length; i++) prefab[i] = new BitWidth(i + 1);
		}
	}
}
